lib/containers/runtime.pm

```diff
@@ -0,0 +1,143 @@
+C<dockerfile_path> is the directory with the Dockerfile.
+C<container_tag> will be the name of the container.
+
+=cut
+sub build {
+    my ($self, $dockerfile_path, $container_tag) = @_;
+    die 'wrong number of arguments' if @_ < 3;
+    assert_script_run "cd $dockerfile_path";
+    $self->_rt_assert_script_run("build -t $container_tag .");
+    record_info "$container_tag created";
+}
+
+=head2 up
+
+Run a container.
+C<image_name> is required and can be the image id, the name or name with tag.
+If C<daemon> is enabled then container will run in the detached mode. Otherwise will be in the 
+interactive mode.
+if C<cmd> found then it will execute the given command into the container.
+The container is always removed after exit.
+
+=cut
+sub up {
+    my ($self, $image_name, %args) = @_;
+    die 'image name or id is required' unless $image_name;
+    my $mode   = $args{daemon} ? '-d'    : '-it';
+    my $remote = $args{cmd}    ? 'sh -c' : '';
+    my $ret    = $self->_rt_script_run(sprintf qq(run --rm %s %s %s '%s'), $mode, $image_name, $remote, $args{cmd});
+    record_info "Remote run on $image_name", $args{cmd};
+    return $ret;
+}
+
+=head2 pull
+
+Pull a container with the given C<tag>.
+
+=cut
+sub pull {
+    my ($self, $tag) = @_;
+    my $ret = $self->_rt_script_run("pull $tag");
+    return $ret;
+}
+
+=head2 enum_images
+
+Return an array ref of the images
+
+=cut
+sub enum_images {
+    my ($self) = shift;
+    my $images_s = $self->_rt_script_output("images -q");
+    record_info "Images", $images_s;
+    my @images = split /[\n\t]/, $images_s;
+    return \@images;
+}
+
+=head2 enum_images
+
+Return an array ref of the containers
+
+=cut
+sub enum_containers {
+    my ($self) = shift;
+    my $containers_s = $self->_rt_script_output("container ls -q");
+    record_info "Containers", $containers_s;
+    my @containers = split /[\n\t]/, $containers_s;
+    return \@containers;
+}
+
+=head2 info
+
+Assert a C<property> against given expected C<value>.
+
+=cut
+sub info {
+    my ($self, %args) = shift;
+    my $property = $args{property} ? qq(--format '{{.$args{property}}}') : '';
+    my $expected = $args{value}    ? qq( | grep $args{value})            : '';
+    $self->_rt_assert_script_run(sprintf("info %s %s", $property, $expected));
+}
+
+=head2 remove_image
+
+Remove a image from the pool.
+
+=cut
+sub remove_image {
+    my ($self, $tag) = @_;
+    $self->_rt_assert_script_run("rmi -f $tag");
+}
+
+=head2 cleanup_system
+
+Remove containers and then all the images respectively and then make sure that everything was cleaned up.
+
+=cut
+sub cleanup_system {
+    my ($self) = shift;
+    # copy from common > clean_container_host
+    $self->_rt_assert_script_run("stop \$($self->{runtime} ps -q)", 180) if script_output("$self->{runtime} ps -q | wc -l") != '0';
+    $self->_rt_assert_script_run("system prune -a -f",              180);
+    assert_equals(0, scalar @{$self->enum_containers()}, "containers have not been removed");
```

Comment:
will it throw error here ? I am not if it is good idea to fail test case if there are some issues with cleanup ...

Reply:
why not. Obviously it is not coming from the images that we test but it is kinda regression issue if the system was not cleaned up.

Reply:
**test isolation** - if I get it right we speaking about cleanup in the host , right ? so imagine there is some bug in Ubuntu which not allow you to cleanup properly . Do you think it would be correct to have red tests which testing SLE Containers when there is some docker bug in Ubuntu/Centos  ?

Reply:
you are correct but i am thinking two things here.
1. the module and the task as now it is scheduled only on sle.
2. For me it is still bug if the docker/podman does not do what it is supposed to do, whatever the system.

As we said the same exact commands are in `containers::common::clean_container_host`. What s the point to run something if you do not verify the outcome in testing.